### PR TITLE
LKE-446: Bootstrap new ci

### PR DESCRIPTION
**Pull request by enys**

### CODE REVIEW

**Comment by enys** [comment]
Merging this commit will imply moving away from the old ci to the new CI

**Comment by enys** [comment]
@davidrapin @francescoinfante @maxime-allex  how would you like to move forward on this one ?

**Comment by davidrapin** [comment]
@enys 
- clarify release needs with backend team
  - do we sometimes need to release for older branches (e.g. 2.8.x?) 
  - ?
- clarify needs for development version
  - how to get rid of dependencies from other projects to the github repo of the rest client directly?) => use `~` or `^` prefix in dev, prevent these prefixes in prod.
  - ?
- merge this PR

thoughts?

**Comment by enys** [comment]
This would make sense for linkurious-shared.
Here the CI does not push an asset. So it is not needed.

**Comment by davidrapin** [comment]
@enys I think the CI should push assets for this repo (i.e. publish to npm)

**Comment by enys** [comment]
I wasn't precise. Old co didn't push any assets at all.
New ci already is configured to publish to npm

**Comment by davidrapin** [comment]
@enys Since this one pushes to NPM, I think we will have to discuss what we discussed yesterday (how to handle dependencies to non-public dev versions) with developers.

Which implies:
- agree on how to handle development builds in this lib (dependency to github vs. dependency to a tag in nexus)
- implement the required changes in this lib + its dependencies (e.g. update the frontend)

Do you want to merge this PR before going forward with this?

**Comment by francescoinfante** [comment]
before merging anything, we need a presentation of the new CI :D

File: package.json [hunk 1]
```diff
@@ -20,8 +20,12 @@
   "scripts": {
     "prepare": "tsc",
     "build": "tsc",
+    "test": "npm run compile && mocha dist/test",
+    "test:unit": "nyc --reporter=text-summary --reporter=cobertura --report-dir=reports  mocha --reporter mocha-junit-reporter --reporter-options mochaFile=./reports/unit/junit-test-results.xml dist/test",
+    "compile": "tsc -b",
```
Comment (francescoinfante) [inline package.json:25]: can you explain the build flag?

File: .gitignore [hunk 2]
```diff
@@ -3,6 +3,11 @@ node_modules
 
 # IDEA config
 .idea
+*.swp
```
Comment (francescoinfante) [inline .gitignore:6]: what is it?

**Comment by davidrapin** [comment]
@enys I think it's time to:
- document the goals (e.g. make PRs safer to merge, run the linter on each commit, only depend on built assets, never release manually), etc.
- document the "how"
- give a short presentation with slides mayb?

**Comment by enys** [comment]
@davidrapin yes !

File: package.json (hunk 1 above)
Comment (enys) [reply package.json:25]: rather than my bad explanation : https://www.typescriptlang.org/docs/handbook/project-references.html

TLDR: faster ( incremental builds)

File: .gitignore (hunk 2 above)
Comment (enys) [reply .gitignore:6]: this is a gobal discussion to have. Either remove all editor files, or have them in each project, with a templated .gitignore.

Specificaly vim here.

**Comment by enys** [comment]
> Do you want to merge this PR before going forward with this?

I think it should be the maintainers choice. However : there is no impact on current workflow, so I would merge.

File: package.json (hunk 1 above)
Comment (francescoinfante) [reply package.json:25]: we already get incremental builds by tsconfig

File: .gitignore (hunk 2 above)
Comment (francescoinfante) [reply .gitignore:6]: just add a # VIM comment for now :D

Comment (enys) [reply .gitignore:6]: a bit like idea config for idea .... :D

File: package.json (hunk 1 above)
Comment (enys) [reply package.json:25]: which isn't present on all projects sadly

Comment (enys) [reply package.json:25]: it is not limited to incremental builds.

**Comment by enys** [comment]
@francescoinfante @davidrapin @aminerhanemi all good to go on this one

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+@Library('linkurious-shared')_
```
Comment (francescoinfante) [inline Jenkinsfile:1]: @enys what does this mean?
rest-client doesn't depend on shared and shared is a private repo :)

**Comment by enys** [comment]
This is not related with nodeJs. This is loading the shared libraries for
Jenkins. Repository named Jenkins-shared-libraries

File: Jenkinsfile (hunk 3 above)
Comment (enys) [reply Jenkinsfile:1]: https://www.jenkins.io/doc/book/pipeline/shared-libraries/

Comment (francescoinfante) [reply Jenkinsfile:1]: calling it linkurious-shared is going to confuse everyone, can we rename it to linkurious-jenkins-shared? (or linkurious-jenkins-common)

Comment (enys) [reply Jenkinsfile:1]: I don't think we should prefix, in a specific context. ie would it make sense to call linkurious-shared linkurious-shared-node ? Here you are specificaly in a Jenkinsfile, It does not look at all like a nodeJs import.

**Comment by enys** [comment]
Lets merge this today, before I have to rebase for the 4th time :)
